### ISSUE TITLE
[WIP]Create v2 release to guide users to a stable "checkpoint" on v2?

### ISSUE THREAD

**Issue by Laurae2** [Microsoft/LightGBM#463]
@guolinke We could create a release as v2 using Microsoft/LightGBM@b6c973a, and the current version we would work on would be 2.1 (or v3?) on master branch. GPU support still need testing and b6c973a (and before) was heavily tested by new kagglers (without raising any big issues).

v2 contains mostly:

* your large v2 PR
* Shrinkage per tree
* Boost from average for regression
* Feature Importance plots
* Large R code changes and support for new objectives
* Restart LightGBM package from scratch in R to clean errors
* Some Python fixes
* Parallel learner fixes
* To add: RMSE metric addition
* To add: MSE metric computation fix

Users in that case would know which version to use if they want something that work out of the box and is stable, but also want to use v2 instead of v1.

Preferably a commit before GPU version, we can add PRs to add all the extra fixes which are known to not break anything:

* python-package: refine plot_tree: https://github.com/Microsoft/LightGBM/pull/388
* python-package: add graphviz.Digraph parameters: https://github.com/Microsoft/LightGBM/pull/400
* Switch RMSE to MSE (true L2 loss): https://github.com/Microsoft/LightGBM/pull/408

**Comment by guolinke** [comment]
sounds good. I will create a new release.

**Comment by guolinke** [comment]
just create the new branch ```v2.0```. We need to add some fixes on that branch:

To-do:
https://github.com/Microsoft/LightGBM/commit/bcb9613ec38220cc894a28e7161b8ec8b3b96aa3
https://github.com/Microsoft/LightGBM/commit/001fe53694c82b711466e9dfb00fb15c1d051798
https://github.com/Microsoft/LightGBM/commit/05f4a26aed5cb7eecfd5a48abbfdc690da579619
https://github.com/Microsoft/LightGBM/commit/60c77487c15a69c9451be80fa9b15e987559a995

**Comment by guolinke** [comment]
@Laurae2 I had merged these specific fixes. 
any other fixes are necessary ?

**Comment by Laurae2** [comment]
I don't think we have any other fixes to add.

Perhaps @wxchan knows if he has other fixes for Python to add.

**Comment by guolinke** [comment]
ok. paste the commit hash here if you think the fix is need. 
@Laurae2 can you also help the test of v2.0 branch ?

**Comment by Laurae2** [comment]
@guolinke For Python package, I test using Travis with #464 .

We need to revert fd0641b both on master, v1, and v2 branch because R does not have documentation anymore.

**Comment by Laurae2** [comment]
Edit: it is commit 0150a1c, not the other one to revert.

**Comment by guolinke** [comment]
@Laurae2 
The Rd file generated by `Roxygen` is gpl license. So we cannot include them ...

**Comment by Laurae2** [comment]
@guolinke Roxygen is an exception. It cannot contaminate `Rd` generated files, because under GPL, Roxygen is similar to an interpreter, and we feed "data" to translate. Such data is not contaminated by GPL as long as you don't bind to it (our Rd files do not require to bind to roxygen at anytime to be usable) or if the license includes a restrictive addition for that specific task (it doesn't).

This is also identical for knitr / Rmarkdown reports: the license will depend on what you have inside, and not on knitr / Rmarkdown itself (for instance, HTML generated files from knitr are licensable with MIT license without any issue, even if knitr is GPL-type).

If this automatic permission was not allowed, most packages non-GPL compliant would be deleted from CRAN (MIT license, Apache, etc.), as CRAN requires all files of a package to be under the same license without any exception (even for licensed datasets).

From https://www.gnu.org/licenses/gpl-faq.en.html#IfInterpreterIsGPL :

> If a programming language interpreter is released under the GPL, does that mean programs written to be interpreted by it must be under GPL-compatible licenses? (#IfInterpreterIsGPL)
>
> When the interpreter just interprets a language, the answer is no. The interpreted program, to the interpreter, is just data; a free software license like the GPL, based on copyright law, cannot limit what data you use the interpreter on. You can run it on any data (interpreted program), any way you like, and there are no requirements about licensing that data to anyone.

**Comment by guolinke** [comment]
@Laurae2 
Our legal team let me remove them. 
I think we can add the command that generate the documents in the R's ReadME.

**Comment by guolinke** [comment]
@wxchan @henry0312 
Do you have another commits that want to add into 2.0 release ?

**Comment by henry0312** [comment]
@guolinke Is https://github.com/Microsoft/LightGBM/pull/457 included?

**Comment by wxchan** [comment]
I think #465 can be added, it's a bug fix. No else.

**Comment by Laurae2** [comment]
@guolinke any way we can add Rd docs without roxygen if your legal team refuses files from roxygen even when the files created can be MIT licensed?

**Comment by guolinke** [comment]
@Laurae2 
Is there any website that can host the rd files? like the readthedocs ? 
So that we can generate separate docs and host them.

**Comment by Laurae2** [comment]
@guolinke I am currently creating a website for LightGBM / xgboost for R about that since last week hosted on Google Sites, and will add much more detailed docs there. But if we can use readthedocs it would be better because documentation would be updated in real time with the commits on this repo.

![image](https://cloud.githubusercontent.com/assets/9083669/25527088/38e41042-2c17-11e7-8bde-68eaa723eb2d.png)

If we stick to my Google Sites, one would need to provide me all the changes everytime (or I can provide specific users a direct access to edit it).

**Comment by guolinke** [comment]
@Laurae2 That website is very awesome !

I am not sure how ReadTheDocs work. Get the markdown file from GitHub? 
 
Welcome to update these rich descriptions back to the https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.md .

**Comment by Laurae2** [comment]
@guolinke I have no idea how to use readthedocs also, I never understood the documentation as it has no clear examples.

It would still be great to have Rd files in the R package, because I know people who are not allowed to use Internet at work when doing data analysis and machine learning for security purposes (Samsung, etc.).

I will put the description in a detailed description markdown file to avoid putting too much text in the parameters file.

**Comment by kmike** [comment]
Would you mind adding fixes related to https://github.com/Microsoft/LightGBM/issues/454 to the 0.2 release? They can be seen as bug fixes for the new "Boost from average for regression" feature.
